baseapp/baseapp.go

```diff
@@ -890,33 +952,31 @@ func (app *BaseApp) DefaultPrepareProposal() sdk.PrepareProposalHandler {
 			byteCount int64
 		)
 
-		iterator := app.mempool.Select(ctx, req.Txs)
+		iterator := h.mempool.Select(ctx, req.Txs)
 		for iterator != nil {
 			memTx := iterator.Tx()
 
-			bz, err := app.txEncoder(memTx)
-			if err != nil {
-				panic(err)
-			}
-
+			bz, err := h.txVerifier.PrepareProposalVerifyTx(memTx)
 			txSize := int64(len(bz))
+			switch {
+			case bz == nil && errors.Is(err, ErrTxEncode):
+				panic(err)
 
-			// NOTE: Since runTx was already executed in CheckTx, which calls
-			// mempool.Insert, ideally everything in the pool should be valid. But
-			// some mempool implementations may insert invalid txs, so we check again.
-			_, _, _, _, err = app.runTx(runTxPrepareProposal, bz)
-			if err != nil {
-				err := app.mempool.Remove(memTx)
+			case bz != nil && err != nil:
+				err := h.mempool.Remove(memTx)
 				if err != nil && !errors.Is(err, mempool.ErrTxNotFound) {
 					panic(err)
 				}
 
 				iterator = iterator.Next()
 				continue
-			} else if byteCount += txSize; byteCount <= req.MaxTxBytes {
-				txsBytes = append(txsBytes, bz)
-			} else {
-				break
+
+			default:
+				if byteCount += txSize; byteCount <= req.MaxTxBytes {
+					txsBytes = append(txsBytes, bz)
+				} else {
+					break
+				}
```

Comment:
Does this `break` exit the `switch` or the `for`?

Reply:
The `break` is in a switch so it breaks the switch. It needs to break the for loop. Good catch. Reverting to original behavior.